### PR TITLE
Allow creation of clusters larger than 500 nodes in GCE, take 2.

### PR DESCRIPTION
**Pull request by gmarek**
The change from the take 1 is to remove the suffix from the name of the
last MIG.

cc @wojtek-t @zmerlynn @brendandburns

### CODE REVIEW

**Comment by gmarek** [comment]
```
• [SLOW TEST:1127.251 seconds]
[Autoscaling] [Skipped]
/usr/local/google/home/gmarek/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/cluster_size_autoscaling.go:105
  Should scale cluster size based on memory reservation
  /usr/local/google/home/gmarek/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/cluster_size_autoscaling.go:104
```

File: cluster/gce/util.sh [hunk 1]
```diff
@@ -810,17 +853,20 @@ function kube-up {
...
 
   echo "Bringing down cluster"
   set +e  # Do not stop on error
 
-  # Delete autoscaler for nodes if present.
+  # Delete autoscaler for nodes if present. We assume that all or none instance groups have an autoscaler
   local autoscaler
   autoscaler=( $(gcloud compute instance-groups managed list --zone "${ZONE}" --project "${PROJECT}" \
-                 | grep "${NODE_INSTANCE_PREFIX}-group" \
+                 | grep "${NODE_INSTANCE_PREFIX}-group-1" \
```
Comment (roberthbailey) [inline cluster/gce/util.sh]: a -1 group may not exist (with less than 500 nodes it won't) so that should check "${NODE_INSTANCE_PREFIX}-group"

**Comment by roberthbailey** [comment]
This looks ok to me, but I'd like @zmerlynn to take a look as well since he found the issue with the last version of this PR.

**Comment by roberthbailey** [comment]
(sorry, didn't mean to close)

File: cluster/gce/util.sh (hunk 1 above)
Comment (gmarek) [inline cluster/gce/util.sh]: Thanks for the catch.

File: cluster/gce/util.sh [hunk 2]
```diff
@@ -713,17 +725,43 @@ function kube-up {
...
+  if [[ ${defaulted_max_instances_per_mig} -le "0" ]]; then
+    echo "MAX_INSTANCES_PER_MIG cannot be negative. Assuming default 500"
+    defaulted_max_instances_per_mig=500
+  fi
+  local num_migs=$(((${NUM_NODES} + ${defaulted_max_instances_per_mig} - 1) / ${defaulted_max_instances_per_mig}))
+  local instances_per_mig=$(((${NUM_NODES} + ${num_migs} - 1) / ${num_migs}))
+  local last_mig_size=$((${NUM_NODES} - (${num_migs} - 1) * ${instances_per_mig}))
+
+  #TODO: parallelize this loop to speed up the process
+  for i in $(seq $((${num_migs} - 1))); do
```
Comment (fgrzadkowski) [inline cluster/gce/util.sh]: Keeping separate names for the main and additional groups is more like a hack. Can we at least have a TODO to fix this at some point?

Comment (wojtek-t) [inline cluster/gce/util.sh]: This is for backward compatibility. We can't easily fix this in one shot.

File: cluster/gce/util.sh [hunk 3]
```diff
@@ -742,9 +780,14 @@ function kube-up {
     METRICS+="--custom-metric-utilization metric=custom.cloudmonitoring.googleapis.com/kubernetes.io/memory/node_reservation,"
     METRICS+="utilization-target=${TARGET_NODE_UTILIZATION},utilization-target-type=GAUGE "
 
-    echo "Creating node autoscaler."
+    echo "Creating node autoscalers."
+
+    for i in $(seq $((${num_migs} - 1))); do
+      gcloud compute instance-groups managed set-autoscaling "${NODE_INSTANCE_PREFIX}-group-$i" --zone "${ZONE}" --project "${PROJECT}" \
+          --min-num-replicas "${AUTOSCALER_MIN_NODES}" --max-num-replicas "${AUTOSCALER_MAX_NODES}" ${METRICS} || true
```
Comment (fgrzadkowski) [inline cluster/gce/util.sh]: I believe we need to adapt MIN_NODES and MAX_NODES accordingly to the number of migs.

File: cluster/gce/util.sh (hunk 2 above)
Comment (gmarek) [inline cluster/gce/util.sh]: No harm in adding a TODO:)

File: cluster/gce/util.sh (hunk 3 above)
Comment (gmarek) [inline cluster/gce/util.sh]: Good catch. Done.

**Comment by wojtek-t** [comment]
@gmarek - once this PR is merged - we should enable "kubernetes-e2e-gce-enormous-cluster" suite on our Jenkins

**Comment by gmarek** [comment]
Yup, I'm just testing it locally - @fgrzadkowski PTAL.

**Comment by piosz** [comment]
Merged manually to unblock scalability effort.

**Comment by ixdy** [comment]
This PR seems to have broken calling `kube-down` when a cluster is not running.

```
2015/12/15 00:48:32 e2e.go:303: Running: teardown
Project: kubernetes-jenkins
Zone: us-central1-f
Shutting down test cluster in background.
ERROR: (gcloud.compute.firewall-rules.delete) Some requests did not succeed:
 - The resource 'projects/kubernetes-jenkins/global/firewalls/e2e-test-parallel-minion-e2e-test-parallel-http-alt' was not found

ERROR: (gcloud.compute.firewall-rules.delete) Some requests did not succeed:
 - The resource 'projects/kubernetes-jenkins/global/firewalls/e2e-test-parallel-minion-e2e-test-parallel-nodeports' was not found

Bringing down cluster using provider: gce
Project: kubernetes-jenkins
Zone: us-central1-f
2015/12/15 00:48:37 e2e.go:309: Error running teardown: exit status 1
2015/12/15 00:48:37 e2e.go:305: Step 'teardown' finished in 4.466300881s
```

Note that we see the output from `detect-project`, but not from `detect-node-names`.
It seems that this all works fine when calling `kube-down` on a running cluster. I don't understand why yet.

@davidopp

**Comment by wojtek-t** [comment]
@gmarek can you please take a look?

**Pull request merged**